### PR TITLE
Fix radix sort test

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

Fix the compilation errors in HeterogeneousCore/CUDAUtilities/test/radixSort_t.cu:
  - take into account the actual size of the type being sorted, instead of always assuming a 64-bit type;
  - use an union instead of a cast to keep the compiler happy about the aliasing rules.

#### PR validation:

`gpuRadixSort_t` compiles and runs fine.

### CODE REVIEW

**Comment by fwyzard** [comment]
type bugfix

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test for CMSSW_12_1_X/slc7_amd64_gcc11 with cms-externals/eigen-git-mirror#7

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
Fixes #34917 .

**Comment by perrotta** [comment]
+1

**Comment by fwyzard** [comment]
The `slc7_amd64_gcc11` test (where this was actually broken) is still running...

**Comment by perrotta** [comment]
> The `slc7_amd64_gcc11` test (where this was actually broken) is [still running](https://cmssdt.cern.ch/jenkins/job/ib-run-pr-tests/17853/console)...

Ah, sorry Andrea. I saw the message that the tests ended succesfully (together with your "+1" for the heterogeneous category) and I thought they were the ones.
Ok, let see how does the relevant test end up, and stay ready to revert in case of issues

**Comment by fwyzard** [comment]
No problem, testing locally it worked, so I'm optimistic :-)

**Comment by fwyzard** [comment]
OK, there are other (preexisting, I assume) errors in `CondCore/EcalPlugins`, `FastSimulation/TrackingRecHitProducer` and `L1Trigger/DTTrigger` - but indeed `HeterogeneousCore/CUDAUtilities` looks good.

File: HeterogeneousCore/CUDAUtilities/test/radixSort_t.cu [hunk 1]
```diff
@@ -32,6 +34,50 @@ struct RS<float> {
...
+// Keep only the `N` most significant bytes of `t`, and set the others to zero
+template <int N, typename T, typename SFINAE = std::enable_if_t<N <= sizeof(T)>>
+void truncate(T& t) {
+  const int shift = 8 * (sizeof(T) - N);
+  union {
+    T t;
+    uintT_t<T> u;
+  } c;
+  c.t = t;
+  c.u = c.u >> shift << shift;
```
Comment (Dr15Jones) [inline HeterogeneousCore/CUDAUtilities/test/radixSort_t.cu:77]: I'm pretty sure this is actually undefined behavior as you are not supposed to set a union using one type and read the union using another type. I believe the recommend way to do this is to use `memcpy`.
```C++
uintT_t<T> u;
memcpy(&u, &t, sizeof(u));
u = u >>shift <<shift;
memcpy(&t, &u, sizeof(t));
```

I played around with that recently on godbolt (for reasons other than this PR) and found compilers know about memcpy and optimize out the calls and just do the 'right' thing.

Comment (Dr15Jones) [reply HeterogeneousCore/CUDAUtilities/test/radixSort_t.cu:77]: I think section 11.5.6.3 of the C++ 20 standard demonstrates that it is undefined behavior.

Comment (makortel) [reply HeterogeneousCore/CUDAUtilities/test/radixSort_t.cu:77]: I'm pretty sure the union pattern is used elsewhere in CMSSW too, so we should address those as well.

**Comment by fwyzard** [comment]
IMHO the C++ standard has kicked itself in a corner with the strict
aliasing rules, and is now clutching at straws.

Personally, I'm fine with undefined behaviour, as long as the compiler does
what I want it to do without complaining.

If you want to propose an implementation based on memcpy and memset, go
ahead - it may even simplify the code avoiding the shifts (but please add
an assert on the endianness of the machine).

.A

**Comment by Dr15Jones** [comment]
Looks like C++ 20 is finally trying to address this

https://en.cppreference.com/w/cpp/numeric/bit_cast

**Comment by fwyzard** [comment]
I've done some tests using `memcpy` instead a `union`, or using `memset`:
```c++
template <int N, typename T, typename SFINAE = std::enable_if_t<N <= sizeof(T)>>
void truncate1(T& t) {
    const int shift = 8 * (sizeof(T) - N);
    union {
        T t;
        uintT_t<T> u;
    } c;
    c.t = t;
    c.u = c.u >> shift << shift;
    t = c.t;
}
```
```c++
template <int N, typename T, typename SFINAE = std::enable_if_t<N <= sizeof(T)>>
void truncate2(T& t) {
    const int shift = 8 * (sizeof(T) - N);
    uintT_t<T> u;
    std::memcpy(&u, &t, sizeof(T));
    u = u >> shift << shift;
    std::memcpy(&t, &u, sizeof(T));
}
```

```c++
template <int N, typename T, typename SFINAE = std::enable_if_t<N <= sizeof(T)>>
void truncate3(T& t) {
    char* bytes = reinterpret_cast<char*>(&t);
    static_assert(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__, "Only little endian architectures are supported");
    std::memset(bytes, 0x00, sizeof(T) - N);
}
```

Host compilers (GCC 11.2, clang 12.0) always generate the same code for `truncate1` and `truncate2`:
  - GCC always generates an `and` with a mask
  - clang generates either an `and` with a mask, or some combinations of move and shifts, or directly sets a byte/word/dword to `0`
 
Device compilers (NVCC 11.3, clang 11.0 `--cuda`) show more variety:
  - both generate an `and` with a mask for `truncate1`
  - NVCC does not optimise out the calls to `memcpy` in `truncate2`, resulting in very verbose and inefficient code
  - clang always generate the same code as for `truncate1` (unlike on the host)

All compilers (host and device) always generate one or more stores of `0` for `truncate3`.

**Comment by fwyzard** [comment]
Summary
  - in host code we could replace the union and assignment with `memcpy` - the result does not change.
  - in device code we should avoid the `memcpy`
  - in all cases, it's not clear to me if `memset` is better or worse than the union

So... keep things as they are for the time being, and revisit once we have C++20 ?

**Comment by fwyzard** [comment]
P.S. my play area on godbolt is here: https://godbolt.org/z/75csPcK1q

**Comment by makortel** [comment]
> So... keep things as they are for the time being, and revisit once we have C++20 ?

I'd be fine with that. Hopefully we get into C++20 before we need to consider other compilers than gcc or clang for pieces of code using union for type punning (that appears to be guaranteed behavior in gcc, I didn't find quickly what clang guarantees).

**Pull request merged**